src/util.rs

```diff
@@ -330,110 +326,66 @@ unsafe impl<T, I: Invariants> TransparentWrapper<I> for Unalign<T> {
-        type Inner = UnsafeCell<$native>;
-
-        // SAFETY: It is "obvious" that each atomic type contains a single
-        // `UnsafeCell` that covers all bytes of the type, but we can also prove
-        // it:
-        // - Since `$atomic` provides an API which permits loading and storing
-        //   values of type `$native` via a `&self` (shared) reference, *some*
-        //   interior mutation must be happening, and interior mutation can only
-        //   happen via `UnsafeCell`. Further, there must be enough bytes in
-        //   `$atomic` covered by an `UnsafeCell` to hold every possible value
-        //   of `$native`.
-        // - Per [1], `$atomic` has the same size as `$native`. This on its own
-        //   isn't enough: it would still be possible for `$atomic` to store
-        //   `$native` using a compact representation (for `$native` types for
-        //   which some bit patterns are illegal). However, this is ruled out by
-        //   the fact that `$atomic` has the same bit validity as `$native` [1].
-        //   Thus, we can conclude that every byte of `$atomic` must be covered
-        //   by an `UnsafeCell`.
-        //
-        // Thus, every byte of `$atomic` is covered by an `UnsafeCell`, and we
-        // set `type Inner = UnsafeCell<$native>`. Thus, `Self` and
-        // `Self::Inner` have `UnsafeCell`s covering the same byte ranges.
-        //
-        // [1] TODO(#896), TODO(https://github.com/rust-lang/rust/pull/121943):
-        //     Cite docs once they've landed.
-        type UnsafeCellVariance = Covariant;
-
-        // SAFETY: No safety justification is required for an invariant
-        // variance.
-        type AlignmentVariance = Invariant;
-
-        // SAFETY: Per [1], all atomic types have the same bit validity as their
-        // native counterparts. The caller has promised that `$atomic` and
-        // `$native` are an atomic type and its native counterpart,
-        // respectively.
-        //
-        // [1] TODO(#896), TODO(https://github.com/rust-lang/rust/pull/121943):
-        //     Cite docs once they've landed.
-        type ValidityVariance = Covariant;
-
-        fn cast_into_inner(ptr: *mut $atomic) -> *mut UnsafeCell<$native> {
-            // SAFETY: Per [1] (from comment on impl block), `$atomic` has the
-            // same size as `$native`. Thus, this cast preserves size.
-            //
-            // This cast trivially preserves provenance.
-            ptr.cast::<UnsafeCell<$native>>()
-        }
-
-        fn cast_from_inner(ptr: *mut UnsafeCell<$native>) -> *mut $atomic {
-            // SAFETY: Per [1] (from comment on impl block), `$atomic` has the
-            // same size as `$native`. Thus, this cast preserves size.
-            //
-            // This cast trivially preserves provenance.
-            ptr.cast::<$atomic>()
-        }
-    };
+/// `Self` must be an atomic type whose native equivalent is `Self::Native`.
```

Comment:
Same here as in `lib.rs` regarding the safety invariant. Ideally this safety invariant should be sufficient to prove that the following impl (`unsafe impl<A: Atomic, I: Invariants> TransparentWrapper<I> for A`) is sound without reference to the stdlib documentation.

I know that I'm asking for a higher quality bar than we have on `main` already 🙂  This refactor is making it so that the safety reasoning is more "at a distance", so I'm a bit more worried about making the safety reasoning precise and watertight.